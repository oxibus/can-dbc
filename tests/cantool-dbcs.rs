#![cfg(feature = "with-serde")]

use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use std::fs::File;
use std::io::Read as _;
use std::path::Path;
use std::{fs, io};

/// Test parsing all DBC files in the `tests/cantools-dbcs` directory, using the `can_dbc` crate.
#[test]
fn test_cantools_dbc_files() {
    // Configure insta snapshot content and location
    with_settings! {
        { omit_expression => true,
          snapshot_path => "snapshots-cantools-dbcs",
          prepend_module_to_snapshot => false },
        {
            test_cantools_dbc_files_int().unwrap();
        }
    }
}

/// Test parsing all DBC files in the `tests/cantools-dbcs` directory.
fn test_cantools_dbc_files_int() -> io::Result<()> {
    for dbc_path in fs::read_dir("./tests/cantools-dbcs")? {
        let path = dbc_path?.path();
        if path
            .extension()
            .and_then(|s| s.to_str())
            .map(|s| s.eq_ignore_ascii_case("dbc"))
            .is_some()
        {
            parse_one_file(&path)?;
        }
    }
    Ok(())
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file(path: &Path) -> io::Result<()> {
    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    eprintln!("Testing DBC file: {file_name}.dbc");
    let mut buffer = Vec::new();
    File::open(path)?.read_to_end(&mut buffer)?;

    match can_dbc::DBC::from_slice(&buffer) {
        Ok(dbc) => assert_yaml_snapshot!(file_name, dbc),
        Err(e) => {
            eprintln!("Failed to parse {file_name}.dbc: {e:?}");
            assert_debug_snapshot!(format!("!error___{file_name}"), e);
        }
    }
    Ok(())
}
