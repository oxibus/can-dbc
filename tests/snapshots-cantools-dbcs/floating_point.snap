---
source: tests/cantool-dbcs.rs
---
version: "HIPBNYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYYY/4/%%%/4/'%**4YYY///"
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - - TestNode
value_tables: []
messages:
  - message_id:
      Standard: 1024
    message_name: Message1
    message_size: 8
    transmitter:
      NodeName: TestNode
    signals:
      - name: Signal1
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 64
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
  - message_id:
      Standard: 1025
    message_name: Message2
    message_size: 8
    transmitter:
      NodeName: TestNode
    signals:
      - name: Signal2
        multiplexer_indicator: Plain
        start_bit: 32
        signal_size: 32
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
      - name: Signal1
        multiplexer_indicator: Plain
        start_bit: 0
        signal_size: 32
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 0
        unit: ""
        receivers:
          - Vector__XXX
message_transmitters: []
environment_variables: []
environment_variable_data: []
signal_types: []
comments:
  - Node:
      node_name: TestNode
      comment: ""
attribute_definitions: []
attribute_defaults: []
attribute_values: []
value_descriptions: []
signal_type_refs: []
signal_groups: []
signal_extended_value_type_list:
  - message_id:
      Standard: 1024
    signal_name: Signal1
    signal_extended_value_type: IEEEdouble64bit
  - message_id:
      Standard: 1025
    signal_name: Signal1
    signal_extended_value_type: IEEEfloat32Bit
  - message_id:
      Standard: 1025
    signal_name: Signal2
    signal_extended_value_type: IEEEfloat32Bit
extended_multiplex: []
